framework/TSA/Wavelet.py

```diff
@@ -0,0 +1,142 @@
+  """
+
+  @classmethod
+  def getInputSpecification(cls):
+    """
+      Method to get a reference to a class that specifies the input data for
+      class cls.
+      @ Out, inputSpecification, InputData.ParameterInput, class to use for
+        specifying input of cls.
+    """
+    import pywt
+    specs = super(Wavelet, cls).getInputSpecification()
+    specs.name = 'Wavelet'
+    specs.descriiption = """TimeSeriesAnalysis algorithm """
+    specs.addSub(InputData.parameterInputFactory(
+      'family',
+      contentType=InputTypes.StringType,
+      descr="""The type of wavelet to use for the transformation. Possible values are:
+      haar, db1, db2, db3, db4, db5, db6,
+      db7, db8, db9, db10, db11, db12, db13, db14, db15, db16, db17, db18, db19,
+      db20, db21, db22, db23, db24, db25, db26, db27, db28, db29, db30, db31, db32,
+      db33, db34, db35, db36, db37, db38 sym2, sym3, sym4, sym5, sym6, sym7, sym8,
+      sym9, sym10, sym11, sym12, sym13, sym14, sym15, sym16, sym17, sym18, sym19,
+      sym20, coif1, coif2, coif3, coif4, coif5, coif6, coif7, coif8, coif9, coif10,
+      coif11, coif12, coif13, coif14, coif15, coif16, coif17, bior1.1, bior1.3,
+      bior1.5, bior2.2, bior2.4, bior2.6, bior2.8, bior3.1, bior3.3, bior3.5,
+      bior3.7, bior3.9, bior4.4, bior5.5, bior6.8 rbio1.1, rbio1.3, rbio1.5,
+      rbio2.2, rbio2.4, rbio2.6, rbio2.8, rbio3.1, rbio3.3, rbio3.5, rbio3.7,
+      rbio3.9, rbio4.4, rbio5.5, rbio6.8, dmey, gaus1, gaus2, gaus3, gaus4, gaus5,
+      gaus6, gaus7, gaus8, mexh, morl, cgau1, cgau2, cgau3, cgau4, cgau5, cgau6,
+      cgau7, cgau8, shan, fbsp, cmor"""
+    ))
+    return specs
+
+
+  def __init__(self, *args, **kwargs):
+    """
+      A constructor that will appropriately intialize a supervised learning object
+      @ In, args, list, an arbitrary list of positional values
+      @ In, kwargs, dict, an arbitrary dictionary of keywords and values
+      @ Out, None
+    """
+    # general infrastructure
+    TimeSeriesAnalyzer.__init__(self, *args, **kwargs)
+
+
+  def handleInput(self, spec):
+    """
+      Reads user inputs into this object.
+      @ In, inp, InputData.InputParams, input specifications
+      @ Out, settings, dict, initialization settings for this algorithm
+    """
+    settings = TimeSeriesAnalyzer.handleInput(self, spec)
+    return settings
+
+
+  def characterize(self, signal, pivot, targets, settings):
+    """
+      This function utilizes the Discrete Wavelet Transform to
+      characterize a time-dependent series of data.
+
+      @ In, signal, np.ndarray, time series with dims [time, target]
+      @ In, pivot, np.1darray, time-like parameter values
+      @ In, targets, list(str), names of targets in same order as signal
+      @ In, settings, dict, additional settings specific to this algorithm
+      @ Out, params, dict, characteristic parameters
+    """
+    # TODO extend to continuous wavelet transform
+    import pywt
```

Comment:
Doesn't this crash if the optional lib isn't installed?

Reply:
This is a great question; I think we've generally checked (at the top of the file) for the optional library, and then conditionally defined the class based on that information, or something. I'm not really an expert on it, @alfoa should be able to provide guidance on how he would like to see it done.